prow/config/tide.go

```diff
@@ -216,3 +248,136 @@ func (tq *TideQuery) Validate() error {
+			option = mergeTideContextPolicy(option, r.Policy)
+			if b, ok := r.Branches[branch]; ok {
+				option = mergeTideContextPolicy(option, b)
+			}
+		}
+	}
+	return option
+}
+
+// GetTideContextPolicy parses the prow config to find context merge options.
+// If none are set, it will use the prow jobs configured and use the default github combined status.
+// Otherwise if set it will use the branch protection setting, or the listed jobs.
+func (c Config) GetTideContextPolicy(org, repo, branch string) (TideContextPolicy, error) {
+	options := parseTideContextPolicyOptions(org, repo, branch, c.Tide.ContextOptions)
+	// Adding required and optional contexts from options
+	required := sets.NewString(options.RequiredContexts...)
+	optional := sets.NewString(options.OptionalContexts...)
+
+	// automatically generate required and optional entries for Prow Jobs
+	prowRequired, prowOptional := BranchRequirements(org, repo, branch, c.Presubmits)
+	required.Insert(prowRequired...)
+	optional.Insert(prowOptional...)
+
+	// Using Branch protection configuration
+	if options.FromBranchProtection != nil && *options.FromBranchProtection {
+		bp, err := c.GetBranchProtection(org, repo, branch)
+		if err != nil {
+			return TideContextPolicy{}, err
+		}
+		if bp == nil {
+			return TideContextPolicy{}, errors.New("branch protection is not set")
+		}
+		if bp.Protect == nil || *bp.Protect {
+			required.Insert(bp.RequiredStatusChecks.Contexts...)
+		}
+	}
+
+	t := TideContextPolicy{
+		RequiredContexts:    required.List(),
+		OptionalContexts:    optional.List(),
+		SkipUnknownContexts: options.SkipUnknownContexts,
+	}
+	if err := t.Validate(); err != nil {
+		return t, err
+	}
+	return t, nil
+}
+
+// IsOptional checks whether a context can be ignored.
+// Will return true if
+// - context is registered as optional
+// - required contexts are registered and the context provided is not required
+// Will return false otherwise. Every context is required.
+func (cp *TideContextPolicy) IsOptional(c string) bool {
+	if sets.NewString(cp.OptionalContexts...).Has(c) {
+		return true
+	}
+	if sets.NewString(cp.RequiredContexts...).Has(c) {
+		return false
+	}
+	if cp.SkipUnknownContexts != nil && *cp.SkipUnknownContexts {
+		return true
+	}
+	return false
+}
+
+// MissingRequiredContexts discard the optional contexts and only look of extra required contexts that are not provided.
+func (cp *TideContextPolicy) MissingRequiredContexts(contexts []string) []string {
+	if len(cp.RequiredContexts) == 0 {
+		return nil
+	}
+	existingContexts := sets.NewString()
+	for _, c := range contexts {
+		existingContexts.Insert(c)
+	}
+	var missingContexts []string
+	for c := range sets.NewString(cp.RequiredContexts...).Difference(existingContexts) {
+		missingContexts = append(missingContexts, c)
+	}
+	return missingContexts
+}
+
+// RegisterOptionalContexts registers optional contexts
+func (cp *TideContextPolicy) RegisterOptionalContexts(c ...string) {
```

Comment:
It might make more sense to just pass the tide context name to GetTideContextPolicy and have that func make the tide context required. That is all that RegisterOptionalContexts is used for now.

Not a big deal either way.

Reply:
We can also just revert part of my change such that tide skip its own status. What do you think ?